src/Nulls.jl

```diff
@@ -115,9 +115,171 @@ xor(::Integer, ::Null) = null
+Base.done(itr::EachReplaceNull, state) = done(itr.x, state)
+Base.eltype(itr::EachReplaceNull) = Nulls.T(eltype(itr.x))
+@inline function Base.next(itr::EachReplaceNull, state)
+    v, s = next(itr.x, state)
+    ((isnull(v) ? itr.replacement : v)::eltype(itr), s)
+end
+
+"""
+    Nulls.skip(itr)
+
+Return an iterator wrapping iterable `itr` which skips [`null`](@ref) values.
+
+Use [`collect`](@ref) to obtain an `Array` containing the non-`null` values in
+`itr`. Note that even if `itr` is a multidimensional array, the result will always
+be a `Vector` since it is not possible to remove nulls while preserving dimensions
+of the input.
+
+See also: [`Nulls.replace`](@ref), [`Nulls.fail`](@ref)
+
+# Examples
+```jldoctest
+julia> collect(Nulls.skip([1, null, 2]))
+2-element Array{Int64,1}:
+ 1
+ 2
+
+julia> collect(Nulls.skip([1 null; 2 null]))
+2-element Array{Int64,1}:
+ 1
+ 2
+
+```
+"""
+skip(itr) = EachSkipNull(itr)
+struct EachSkipNull{T}
+    x::T
+end
+Base.iteratorsize(::Type{<:EachSkipNull}) =
+    Base.SizeUnknown()
+Base.iteratoreltype(::Type{EachSkipNull{T}}) where {T} =
+    Base.iteratoreltype(T)
+Base.eltype(itr::EachSkipNull) = Nulls.T(eltype(itr.x))
+# Fallback implementation for general iterables: we cannot access a value twice,
+# so after finding the next non-null element in start() or next(), we have to
+# pass it in the iterator state, which introduces a type instability since the value
+# is null if the input does not contain any non-null element.
+# As of Julia 0.6 and early 0.7, this instability kills performance.
+@inline function Base.start(itr::EachSkipNull)
+    s = start(itr.x)
+    v = null
+    @inbounds while !done(itr.x, s) && isnull(v)
+        v, s = next(itr.x, s)
+    end
+    (v, s)
+end
+@inline Base.done(itr::EachSkipNull, state) = isnull(state[1]) && done(itr.x, state[2])
+@inline function Base.next(itr::EachSkipNull, state)
+    v1, s = state
+    v2 = null
+    @inbounds while !done(itr.x, s) && isnull(v2)
+        v2, s = next(itr.x, s)
+    end
+    (v1::eltype(itr), (v2, s))
+end
+# Optimized implementation for AbstractArray, relying on the ability to access x[i] twice:
+# once in done() to find the next non-null entry, and once in next() to return it.
+# This works around the type instability problem of the generic fallback.
+@inline function _next_nonnull_ind(x::AbstractArray, s)
+    idx = eachindex(x)
+    @inbounds while !done(idx, s)
+        i, new_s = next(idx, s)
+        isnull(x[i]) || break
+        s = new_s
+    end
+    s
+end
+@inline Base.start(itr::EachSkipNull{<:AbstractArray}) =
+    _next_nonnull_ind(itr.x, start(eachindex(itr.x)))
+@inline Base.done(itr::EachSkipNull{<:AbstractArray}, state) =
+    done(eachindex(itr.x), state)
+@inline function Base.next(itr::EachSkipNull{<:AbstractArray}, state)
+    i, state = next(eachindex(itr.x), state)
+    @inbounds v = itr.x[i]::eltype(itr)
+    (v, _next_nonnull_ind(itr.x, state))
+end
+
+"""
+    Nulls.fail(itr)
```

Comment:
I still don't really understand the use for this iterator/function. Can you remind me?

Reply:
Well, it can be used to be sure you don't pass an iterator containing nulls to a function which would accept it silently. It's faster than `any(isnull, x)` since checking happens on the fly. Currently `sum(Nulls.fail(x))` also much faster than `sum(x)` since there's no type instability, but with compiler improvements I guess this difference could go away.